pkg/vm/vm_core.go

```diff
@@ -325,13 +320,63 @@ func (vm *VirtualMachine) ComputeOperands(instruction Instruction) (Operands, er
 
 	operands := Operands{
 		Dst: *dst,
-		Op0: *op0,
-		Op1: *op1,
+		Op0: op0,
+		Op1: op1,
 		Res: res,
 	}
 	return operands, nil
 }
 
+// Runs deductions for Op0, first runs builtin deductions, if this fails, attempts to deduce it based on dst and op1
+// Also returns res if it was also deduced in the process
+// Inserts the deduced operand
+// Fails if Op0 was not deduced or if an error arised in the process
+func (vm *VirtualMachine) ComputeOp0Deductions(op0_addr memory.Relocatable, instruction *Instruction, dst *memory.MaybeRelocatable, op1 *memory.MaybeRelocatable) (deduced_op0 memory.MaybeRelocatable, deduced_res *memory.MaybeRelocatable, err error) {
+	op0, err := vm.DeduceMemoryCell(op0_addr)
+	if err != nil {
+		return *memory.NewMaybeRelocatableFelt(lambdaworks.FeltZero()), nil, err
+	}
+	if op0 == nil {
+		op0, deduced_res, err = vm.DeduceOp0(instruction, dst, op1)
+		if err != nil {
+			return *memory.NewMaybeRelocatableFelt(lambdaworks.FeltZero()), nil, err
+		}
+	}
+	if op0 != nil {
+		vm.Segments.Memory.Insert(op0_addr, op0)
+	} else {
+		return *memory.NewMaybeRelocatableFelt(lambdaworks.FeltZero()), nil, errors.New("Failed to compute or deduce op0")
+	}
```

Comment:
I don't understand these conditionals, the `else` from line 347 is the same as the condition from line 339, right?

Reply:
The logic would be as follows : 
op0 is deduced (DeducedOp0) now op0 can be some or none
if op0 is none (line 339) deduce it via builtins (DeduceMemoryCell) now op0 can be some or none
Then is op0 is some (line 345), it was deduced by any of the two methods, so we need to insert it into memory
If op0 is still none, an error is returned

Reply:
Makes sense